File: csharp/347_Top_K_Frequent_Elements.cs
public class Solution {
    public int[] TopKFrequent(int[] nums, int k) {
        int[] result = new int[k];
        Dictionary<int, int> m = new Dictionary<int, int>();
        List<int> q = new List<int>();

        // Frequency calculate
        foreach (int num in nums) {
            if (!m.ContainsKey(num)) 
                m.Add(num, 1);
            else
                m[num]++;            
        }

        // sort frequency
        q = m.Keys.ToList();
        q.Sort(delegate(int x, int y) {
            return m[y].CompareTo(m[x]);
        });

        // get k result
        for (int i = 0; i < k; i++) {
            result[i] = q[i];
        }

        return result;
    }
}

File: cpp/347_Top_K_Frequent_Elements.cpp
// O(NLogK)
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> m;
        vector<int> result;
        priority_queue<pair<int, int>> q;

        for (auto num : nums) m[num]++; // calculate frequency
        for (auto num : m) q.push({num.second, num.first}); // use heap to sort frequency
        while (k--) {
            result.push_back(q.top().second);
            q.pop();
        }           

        return result;
    }
};